utils/tfw_logger.cc

```diff
@@ -0,0 +1,463 @@
+} TfwField;
+
+static const TfwField tfw_fields[] = {
+	[TFW_MMAP_LOG_ADDR]		= {"address", clickhouse::Type::IPv6},
+	[TFW_MMAP_LOG_METHOD]		= {"method", clickhouse::Type::UInt8},
+	[TFW_MMAP_LOG_VERSION]		= {"version", clickhouse::Type::UInt8},
+	[TFW_MMAP_LOG_STATUS]		= {"status", clickhouse::Type::UInt16},
+	[TFW_MMAP_LOG_RESP_CONT_LEN]	= {"response_content_length", clickhouse::Type::UInt32},
+	[TFW_MMAP_LOG_RESP_TIME]	= {"response_time", clickhouse::Type::UInt32},
+	[TFW_MMAP_LOG_VHOST]		= {"vhost", clickhouse::Type::String},
+	[TFW_MMAP_LOG_URI]		= {"uri", clickhouse::Type::String},
+	[TFW_MMAP_LOG_REFERER]		= {"referer", clickhouse::Type::String},
+	[TFW_MMAP_LOG_USER_AGENT]	= {"user_agent", clickhouse::Type::String},
+	[TFW_MMAP_LOG_DROPPED]		= {"dropped_events", clickhouse::Type::UInt64},
+};
+
+std::atomic<bool> stop_flag{false};
+
+#ifdef DEBUG
+static void
+dbg_hexdump(const char *data, int buflen)
+{
+	const unsigned char *buf = (const unsigned char*)data;
+	std::ostringstream oss;
+	oss << std::hex << std::setfill('0');
+
+#define PRINT_CHAR(c) (std::isprint(c) ? c : '.')
+	for (int i = 0; i < buflen; i += 16) {
+		oss << std::setw(6) << i << ": ";
+
+		for (int j = 0; j < 16; ++j)
+			if (i + j < buflen)
+				oss << std::setw(2) << (unsigned)buf[i + j] << " ";
+			else
+				oss << "   ";
+		oss << " ";
+		for (int j = 0; j < 16; ++j) {
+			if (i + j >= buflen)
+				break;
+			oss << (char)PRINT_CHAR(buf[i + j]);
+		}
+		oss << std::endl;
+	}
+	oss << std::dec << "len = " << buflen << std::endl;
+	std::cout << oss.str();
+#undef PRINT_CHAR
+}
+#else
+static void
+dbg_hexdump([[maybe_unused]] const char *data, [[maybe_unused]] int buflen)
+{
+}
+#endif /* DEBUG */
+
+static std::shared_ptr<clickhouse::Block>
+make_block()
+{
+	unsigned int i;
+	auto block = std::make_shared<clickhouse::Block>();
+
+	auto col = std::make_shared<clickhouse::ColumnDateTime64>(3);
+	block->AppendColumn("timestamp", col);
+
+	for (i = TFW_MMAP_LOG_ADDR; i < TFW_MMAP_LOG_MAX; ++i) {
+		const TfwField *field = &tfw_fields[i];
+
+		auto col = tfw_column_factory(field->code);
+		block->AppendColumn(field->name, col);
+	}
+
+	return block;
+}
+
+int
+read_access_log_event(const char *data, int size, TfwClickhouse *clickhouse)
+{
+	auto block = clickhouse->get_block();
+	const char *p = data;
+	TfwBinLogEvent *event = (TfwBinLogEvent *)p;
+	int i;
+
+	p += sizeof(TfwBinLogEvent);
+	size -= sizeof(TfwBinLogEvent);
+
+#define INT_CASE(method, col_type, val_type)					\
+	case method:								\
+		if (TFW_MMAP_LOG_FIELD_IS_SET(event, i)) {			\
+			if (len > size)						\
+				return -1;					\
+			(*block)[ind]->As<col_type>()->Append(*(val_type *)p);	\
+		} else								\
+			(*block)[ind]->As<col_type>()->Append(0);		\
+		break;
+
+
+	(*block)[0]->As<clickhouse::ColumnDateTime64>()->Append(event->timestamp);
+
+	for (i = TFW_MMAP_LOG_ADDR; i < TFW_MMAP_LOG_MAX; ++i) {
+		int len, ind = i + 1;
+
+		len = tfw_mmap_log_field_len((TfwBinLogFields)i);
```

Comment:
It seems that this `len` calculation is used only for `INT_CASE`. for other cases we recalculate `len = *((uint16_t *)p);` so may be move `len = tfw_mmap_log_field_len((TfwBinLogFields)i);` to INT_CASE macro?

Reply:
Also we check TFW_MMAP_LOG_FIELD_IS_SET two times at the beginning of loop and at the end of the loop. Why we need second TFW_MMAP_LOG_FIELD_IS_SET check and can't calculate
```
p += len;
size -= len;
```
at first TFW_MMAP_LOG_FIELD_IS_SET checking?

Reply:
Fixed.